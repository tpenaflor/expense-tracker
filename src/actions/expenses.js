
// import {v1} from 'uuid'
import db from '../firebase/firebase'

export const addExpensee = (expense) => ({
    type:"ADD_EXPENSE",
    expense
})

export const addExpense = (expenseData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        const {desc='', note='',amount=0, createAt=0} = expenseData
        const expense = {desc, note, amount, createAt}

        return db.ref(`expenses/${uid}`).push(expense).then((snap) => {
            dispatch(addExpensee({id:snap.key, ...expense}))
        })
    }
}

export const removeExpensee = (id) => ({
    type:"REMOVE_EXPENSE",
    expense : {
        id
    }
})

export const removeExpense = (id) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return db.ref(`expenses/${uid}/${id}`).remove().then(() =>{
            dispatch(removeExpensee(id))
        })
    }
}

export const editExpensee = (id, update) => ({
    type:"EDIT_EXPENSE",
    id,
    update
})

export const editExpense = (id, update) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        return db.ref(`expenses/${uid}/${id}`).update(update).then(() =>{
            dispatch(editExpensee(id, update))
        })
    }
}

export const setExpensess = (expenses) => ({
    type : 'SET_EXPENSES',
    expenses
})

export const setExpenses = (uid) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid

        let expenses = []
        return db.ref(`expenses/${uid}`).once('value').then((snap) =>{
            snap.forEach((expense) => {
                expenses.push({id:expense.key, ...expense.val()})
            });

            dispatch(setExpensess(expenses))
        })
    }
}