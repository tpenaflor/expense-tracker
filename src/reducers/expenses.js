

const expensesReducerDefaultState = []

const expensesReducer = (state = expensesReducerDefaultState, {type, expense, id, update, expenses}) => {
    switch (type) {
        case 'ADD_EXPENSE' :
            return [...state, expense]
        case 'REMOVE_EXPENSE' :
            return state.filter(({id})=> id !== expense.id)
        case 'EDIT_EXPENSE' :
            return state.map((entry) => {
                if (entry.id === id){
                    return {...entry, ...update}
                }
                return entry
            })
        case 'SET_EXPENSES' :
            return expenses
        default:
            return state
    }
}

export default expensesReducer