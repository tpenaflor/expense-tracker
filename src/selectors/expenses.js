
import moment from 'moment'

export const getVisibleExpenses = (expenses, {text, startDate, endDate, sortBy}) => {

    return expenses.filter(({desc, createAt})=> {
        const createMoment = moment(createAt)
        // const startMatch = typeof startDate !== 'number' || createAt >= startDate
        // const endMatch = typeof endDate !== 'number' || createAt <= endDate
        const startMatch = startDate ? startDate.isSameOrBefore(createMoment, 'day'): true
        const endMatch = endDate ? endDate.isSameOrAfter(createMoment, 'day'): true
        const textMatch = desc.toLowerCase().includes(text.toLowerCase())
        
        return startMatch && endMatch && textMatch
    }).sort((a, b)=>{
        if (sortBy === 'date') return a.createAt - b.createAt
        if (sortBy === 'amount') return a.amount - b.amount
    });
}

export const getTotalExpenses = (expenses) => expenses.reduce((total, {amount}) => total += amount, 0)

export default getVisibleExpenses