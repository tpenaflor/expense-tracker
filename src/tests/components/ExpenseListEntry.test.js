import React from 'react'
import {shallow} from 'enzyme'

import ExpenseListEntry from '../../components/ExpenseEntry'
import expenses from '../fixtures/expenses'


test('test expense list entry', ()=>{
    const wrapper = shallow(<ExpenseListEntry {...expenses[1]}/>)
    expect(wrapper).toMatchSnapshot()
})
