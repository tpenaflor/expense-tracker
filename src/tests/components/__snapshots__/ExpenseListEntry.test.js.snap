// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test expense list entry 1`] = `
<div
  className="entry"
>
  <h2>
    <Link
      to="/editExpense/2"
    >
      coffee
    </Link>
  </h2>
  <p>
    $25,000.00 on Dec 01, 1969
  </p>
</div>
`;
